test/e2e/e2e.go

```diff
@@ -40,3 +45,83 @@ func SetupWithNamespace(t *testing.T, namespace string) *test.Clients {
+func deploymentRecreation(t *testing.T, clients *test.Clients, names test.ResourceNames) {
+	dpListSave, err := clients.KubeClient.Kube.AppsV1().Deployments(names.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get any deployment under the namespace %q: %v",
+			test.ServingOperatorNamespace, err)
+	}
+	if len(dpListSave.Items) == 0 {
+		t.Fatalf("No deployment under the namespace %q was found",
+			test.ServingOperatorNamespace)
+	}
+	// Delete the deployments one by one to see if they will be recreated.
+	for _, deployment := range dpListSave.Items {
+		if err := clients.KubeClient.Kube.AppsV1().Deployments(deployment.Namespace).Delete(deployment.Name,
+			&metav1.DeleteOptions{}); err != nil {
+			t.Fatalf("Failed to delete deployment %s/%s: %v", deployment.Namespace, deployment.Name, err)
+		}
+
+		waitErr := wait.PollImmediate(resources.Interval, resources.Timeout, func() (bool, error) {
+			dep, err := clients.KubeClient.Kube.AppsV1().Deployments(deployment.Namespace).Get(deployment.Name, metav1.GetOptions{})
+			if err != nil && apierrs.IsNotFound(err) {
+				// If the deployment is not found, we continue to wait for the availability.
+				return false, nil
+			}
+			return resources.IsDeploymentAvailable(dep)
+		})
+
+		if waitErr != nil {
+			t.Fatalf("The deployment %s/%s failed to reach the desired state: %v", deployment.Namespace, deployment.Name, err)
+		}
+
+		if _, err := resources.WaitForKnativeServingState(clients.KnativeServingAlphaClient, test.ServingOperatorName,
+			resources.IsKnativeServingReady); err != nil {
+			t.Fatalf("KnativeService %q failed to reach the desired state: %v", test.ServingOperatorName, err)
+		}
+		t.Logf("The deployment %s/%s reached the desired state.", deployment.Namespace, deployment.Name)
+	}
+}
+
+// knativeServingDeletion deletes tha KnativeServing to see if all the deployments will be removed.
+func knativeServingDeletion(t *testing.T, clients *test.Clients, names test.ResourceNames) {
+	if err := clients.KnativeServingAlphaClient.Delete(names.KnativeServing, &metav1.DeleteOptions{}); err != nil {
+		t.Fatalf("KnativeService %q failed to delete: %v", names.KnativeServing, err)
+	}
+	if _, err := resources.WaitForKnativeServingState(clients.KnativeServingAlphaClient, names.KnativeServing,
+		resources.IsKnativeServingDeleted); err != nil {
+		t.Fatalf("KnativeService %q failed to be deleted: %v", names.KnativeServing, err)
+	}
+
+	dpListSave, err := clients.KubeClient.Kube.AppsV1().Deployments(names.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("Error getting any deployment under the namespace %q: %v", names.Namespace, err)
+	}
+
+	for _, deployment := range dpListSave.Items {
+		waitErr := wait.PollImmediate(resources.Interval, resources.Timeout, func() (bool, error) {
+			if _, err := clients.KubeClient.Kube.AppsV1().Deployments(deployment.Namespace).Get(deployment.Name,
+				metav1.GetOptions{}); err != nil && apierrs.IsNotFound(err) {
```

Comment:
We're ignoring other errors here, which might cause the tests to fail for bogus reasons. I'd envision the logic to look more like:

```golang
_, err := clients.KubeClient.Kube.AppsV1().Deployments(deployment.Namespace).Get(deployment.Name, metav1.GetOptions{})
if err != nil {
  if apierrs.IsNotFound(err) {
    return true, nil
  }
  return false, err
}
return false, nil
```

That'll cause other errors (like a wrongly setup test env) to be surfaced properly.

Reply:
Done